### PR TITLE
Performance improvements for row to column conversions

### LINKED ISSUE #10054: [FEA] Improve performance of row to column by changing order of processing
**Is your feature request related to a problem? Please describe.**
Row to column conversions are currently about 7 times slower than column to row. It is possible the cause of this is the direction of travel of the threads in the conversion process. Currently the tiles are built going down the rows before moving to the next group of columns. This could be causing access issues when converting from row to column as the data is in row-major format.

**Describe the solution you'd like**
Investigation is necessary to see if this is causing the performance disparity. The tile creation code is currently run twice, once to size the destination array and once to fill it. This can be changed to size the tiles and then build the tiles so there isn't this duplication of work. A benefit of this would be that it would be much easier to generate a different ordering of tiles based on the operation.

**Describe alternatives you've considered**
We could sort the array of tile infos after generation based on the operation.

### LINKED ISSUE #10055: [FEA] Investigate thread striding for row/column conversions
**Is your feature request related to a problem? Please describe.**
Currently the threads stride over the data to copy in the row/column conversion code. Each thread copies one element and then skips over the number of threads and copies another element until they are all copied. This results in each copy having to determine the column size, column offset, and row offset of each element. 

**Describe the solution you'd like**
If the threads worked on vertical stripes of data, they could cache this column information and only fetch the row offset for each element. This would reduce the amount of global reads from 4 to 2 for each element and should improve performance as a result. This could also be the cause of the uncoalesced access that we see in the nsight traces.

**Describe alternatives you've considered**
Each thread could cache the last column's information and only update it when necessary and continue the current striding.

### PR DESCRIPTION
**Pull request by hyperbolic2346**
This is the second-half of the performance changes for row to column conversions. The first half is in PR #11063 and this half completes the adjustments. Good wins here from changing the way data is read and written.

~Leaving this as draft until #11063 is merged so this cleans up to just what is changed for this portion.~

```Comparing branch-22.08/ROW_CONVERSION_BENCH to mwilson/row_to-column_optimization/ROW_CONVERSION_BENCH
Benchmark                                                                               Time             CPU      Time Old      Time New       CPU Old       CPU New
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
RowConversion/old_to_row_conversion/1048576/manual_time                              +0.0020         +0.0020             5             5             5             5
RowConversion/new_to_row_conversion/1048576/manual_time                              -0.5679         -0.5657            16             7            16             7
RowConversion/new_to_row_extended_conversion/1048576/manual_time                     -0.5642         -0.5636            16             7            16             7
RowConversion/string_to_row_extended_conversion/1048576/manual_time                  -0.5734         -0.5743            27            12            27            12
RowConversion/old_from_row_conversion/1048576/manual_time                            -0.0062         -0.0057             4             4             4             4
RowConversion/new_from_row_conversion/1048576/manual_time                            -0.7678         -0.7674            31             7            32             7
RowConversion/new_from_row_extended_conversion/1048576/manual_time                   -0.7680         -0.7675            31             7            32             7
RowConversion/string_from_row_extended_conversion/1048576/manual_time                -0.5559         -0.5667            39            17            40            17
```

The major changes for this performance change revolve around data coalescing. By changing what thread was reading what data and keeping the data a warp is accessing linear in memory, the coalescing was improved significantly here. Specifically:

- All double-buffering was removed. This added complexity to the code and also proved to be a slower approach to dedicating a single block to a tile.
- `copy_from_rows` Changed from using a thread per row to using a warp per row, allowing more threads to participate in the memcpy of the data.
- `copy_validity_from_rows` Changed from a thread copying 8 bytes of validity data and striding to using a warp to read a column of data at a time.

closes #10055 
closes #10054

### CODE REVIEW

**Comment by hyperbolic2346** [comment]
Last part of performance before this is moved to `spark-rapids-jni`

**Comment by ttnghia** [comment]
Wow, up to `70%` reduction in running time. That's great.

**Comment by ttnghia** [comment]
Can you write more description about what helps with performance improvement?

File: java/src/main/native/src/row_conversion.cu [hunk 1]
```diff
@@ -1050,9 +1020,7 @@ copy_validity_from_rows(const size_type num_rows, const size_type num_columns,
                         size_type const *batch_row_boundaries, bitmask_type **output_nm,
                         const size_type validity_offset, device_span<const tile_info> tile_infos,
                         const int8_t *input_data) {
-  extern __shared__ int8_t shared_data[];
-  int8_t *shared_tiles[NUM_VALIDITY_TILES_PER_KERNEL_LOADED] = {
-      shared_data, shared_data + shmem_used_per_tile / 2};
+  extern __shared__ int8_t shared[];
```
Comment (ttnghia) [inline java/src/main/native/src/row_conversion.cu:1023]: What about the same statement at line 915?

Comment (hyperbolic2346) [reply java/src/main/native/src/row_conversion.cu:1023]: Correct, each kernel is using shared memory and must declare it.

File: java/src/main/native/src/row_conversion.cu [hunk 2]
```diff
@@ -923,106 +910,89 @@ __global__ void copy_from_rows(const size_type num_rows, const size_type num_col
   // to speed up some of the random access memory we do, we copy col_sizes and col_offsets
   // to shared memory for each of the tiles that we work on
 
-  constexpr unsigned stages_count = NUM_TILES_PER_KERNEL_LOADED;
   auto group = cooperative_groups::this_thread_block();
-  extern __shared__ int8_t shared_data[];
-  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_tile};
+  auto warp = cooperative_groups::tiled_partition<cudf::detail::warp_size>(group);
```
Comment (ttnghia) [inline java/src/main/native/src/row_conversion.cu:914]: Can this be `constexpr`?

Comment (hyperbolic2346) [reply java/src/main/native/src/row_conversion.cu]: They can't be constexpr, but the can be const. Changed that.

File: java/src/main/native/src/row_conversion.cu [hunk 3]
```diff
@@ -100,31 +94,26 @@ namespace detail {
...
- * Internally, this is referred to as the row batch, which is a group of rows
- * that will fit into this 2 gig space requirement. There are typically 1 of
- * these batches, but there can be 2.
+ * The row data has to fit inside a cuDF column, which limits it to 2 gigs
+ * currently. The calling code attempts to keep the data size under 2 gigs,
+ * but due to padding this isn't always the case, so being able to break
+ * this up into multiple columns is necessary. Internally, this is
+ * referred to as the row batch, which is a group of rows that will fit
+ * into this 2 gig space requirement. There are typically 1 of these
+ * batches, but there can be 2.
```
Comment (ttnghia) [inline java/src/main/native/src/row_conversion.cu:103]: It seems that the comments in this file are line-broken manually. If using clang-format in cudf (format-on-save), you can typically write very long lines then the formatter will break line for you.

File: java/src/main/native/src/row_conversion.cu [hunk 4]
```diff
@@ -923,106 +910,89 @@ __global__ void copy_from_rows(const size_type num_rows, const size_type num_col
...
+
+        auto const shared_memory_offset =
+            col_offsets[absolute_col] - col_offsets[tile.start_col] + shared_memory_row_offset;
+        auto const column_size = col_sizes[absolute_col];
+
+        int8_t *shmem_src = &shared[shared_memory_offset];
+        int8_t *dst = &output_data[absolute_col][absolute_row * column_size];
 
-      MEMCPY(dst, shmem_src, column_size, processing_barrier);
+        MEMCPY(dst, shmem_src, column_size, tile_barrier);
```
Comment (ttnghia) [inline java/src/main/native/src/row_conversion.cu:988]: I usually see the `memcpy` APIs operating on `void*` pointers. Using `int8_t` is of course still valid, but I'm not sure if there will be any conversion to `void*` under the hood? If so, the compiler may issue warnings.

File: java/src/main/native/src/row_conversion.cu [hunk 5]
```diff
@@ -1050,136 +1020,102 @@ copy_validity_from_rows(const size_type num_rows, const size_type num_columns,
...
+    auto const relative_row = section_y * rows_per_read + warp.thread_rank();
+    auto const absolute_col = relative_col + tile_start_col;
+    auto const absolute_row = relative_row + tile_start_row;
+    auto const row_batch_start =
+        tile.batch_number == 0 ? 0 : batch_row_boundaries[tile.batch_number];
+
+    auto const participation_mask = __ballot_sync(0xFFFFFFFF, absolute_row < num_rows);
+
+    if (absolute_row < num_rows) {
+      auto const my_byte = input_data[row_offsets(absolute_row, row_batch_start) + validity_offset +
```
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply java/src/main/native/src/row_conversion.cu:1074]: A little ascii art here might go a long way. Ala:
```
      C0  C1  C2  C3  C4  C5  C6  C7
R0    1   0   1   0   0   1   1   0            <-- thread 0 reads byte r0
R1    1   1   1   1   1   1   1   0            <-- thread 1 reads byte r1
R2    0   0   1   0   0   1   1   0            <-- thread 2 reads byte r2
...
R31   1   1   1   1   1   1   1   1

      ^ 
      |  1 bit of each input byte, by column, are swizzled into a single 32 bit word via __ballot_sync, representing 
         32 rows of that column.  
```
Something like that.

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

File: java/src/main/native/src/row_conversion.cu (hunk 4 above)
Comment (mythrocks) [reply java/src/main/native/src/row_conversion.cu:988]: Pardon me, but `T*` to `void*` [is a ~standard~ *implicit* conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion). Why would the compiler complain?

Comment (ttnghia) [reply java/src/main/native/src/row_conversion.cu:988]: They may complain about some implicit conversions. I'm not sure if the compilers will complain about this case.

Comment (mythrocks) [reply java/src/main/native/src/row_conversion.cu:988]: I'd recommend an explicit cast only if the compiler complains. (And that would be surprising.)

File: java/src/main/native/src/row_conversion.cu [hunk 6]
```diff
@@ -593,8 +581,8 @@ __global__ void copy_to_rows(const size_type num_rows, const size_type num_colum
   // This has been broken up for us in the tile_info struct, so we don't have
   // any calculation to do here, but it is important to note.
 
-  auto group = cooperative_groups::this_thread_block();
-  auto warp = cooperative_groups::tiled_partition<cudf::detail::warp_size>(group);
+  auto const group = cooperative_groups::this_thread_block();
```
Comment (mythrocks) [inline java/src/main/native/src/row_conversion.cu:584]: :+1:

**Comment by mythrocks** [comment]
> Changed from using a thread per row to using *a warp per row*, allowing more threads to participate in the memcpy of the data.

In `convert_from_rows()`, are we using a warp per tile, rather than a warp per row?

File: java/src/main/native/src/row_conversion.cu [hunk 7]
```diff
@@ -923,106 +910,89 @@ __global__ void copy_from_rows(const size_type num_rows, const size_type num_col
...
+    auto const fetch_tile_row_size = fetch_tile.get_shared_row_size(col_offsets, col_sizes);
+    auto const row_batch_start =
+        fetch_tile.batch_number == 0 ? 0 : batch_row_boundaries[fetch_tile.batch_number];
+
+    for (int absolute_row = warp.meta_group_rank() + fetch_tile.start_row;
+         absolute_row <= fetch_tile.end_row; absolute_row += warp.meta_group_size()) {
+      warp.sync();
+      auto shared_offset = (absolute_row - fetch_tile_start_row) * fetch_tile_row_size;
+      auto dst = &shared[shared_offset];
+      auto src = &input_data[row_offsets(absolute_row, row_batch_start) + starting_col_offset];
```
Comment (mythrocks) [inline java/src/main/native/src/row_conversion.cu:939]: Hah! I just realized that `RowOffsetFunctor` might be a more appropriate name for `RowOffsetIter`.

File: java/src/main/native/src/row_conversion.cu [hunk 8]
```diff
@@ -1050,136 +1020,102 @@ copy_validity_from_rows(const size_type num_rows, const size_type num_columns,
...
+    auto const participation_mask = __ballot_sync(0xFFFFFFFF, absolute_row < num_rows);
+
+    if (absolute_row < num_rows) {
+      auto const my_byte = input_data[row_offsets(absolute_row, row_batch_start) + validity_offset +
+                                      absolute_col / cols_per_read];
+
+      // so every thread that is participating in the warp has a byte, but it's row-based
+      // data and we need it in column-based. So we shuffle the bits around to make
+      // the bytes we actually write.
+      for (int i = 0, byte_mask = 0x1; i < cols_per_read && relative_col + i < num_columns;
```
Comment (mythrocks) [inline java/src/main/native/src/row_conversion.cu:1080]:
```suggestion
      for (int i = 0, byte_mask = 0x1; (i < cols_per_read) && ((relative_col + i) < num_columns);
```
Based on operator precedence and associativity, the original code does do the right thing, of course. I'd feel just a little safer if the ordering were explicit.

**Review by mythrocks — CHANGES_REQUESTED** [review]
Some minor nitpicks. I need to get my head around the validity calculation. I'll pore over this again.

**Comment by mythrocks** [comment]
Forgive me, for I've taken the liberty to fix the formatting in the PR's description. Please feel free to delete the strike-through text.

File: java/src/main/native/src/row_conversion.cu [hunk 9]
```diff
@@ -133,10 +122,9 @@ namespace detail {
  *
  * Tiles:
  * The tile info describes a tile of data to process. In a GPU with
- * 48KB of shared memory each tile uses approximately 24KB of memory
- * which equates to about 144 bytes in each direction. The tiles are
- * kept as square as possible to attempt to coalesce memory operations.
- * The taller a tile is the better coalescing of columns, but row
+ * 48KB this equates to about 288 bytes in each direction of a table.
```
[the remark this answers is not in the document]
Comment (hyperbolic2346) [reply java/src/main/native/src/row_conversion.cu:125]: This is correct, I'll update.

**Comment by hyperbolic2346** [comment]
> From the description:
> 
> > `copy_from_rows`: Changed from using a thread per row to using _a warp per row_, allowing more threads to participate in the memcpy of the data.
> 
> In `copy_from_rows()`, are we using a warp per tile, rather than a warp per row?

We march the tile using a warp per row.

File: java/src/main/native/src/row_conversion.cu (hunk 7 above)
Comment (mythrocks) [reply java/src/main/native/src/row_conversion.cu:931]: Although I read it as `Iter", I kept converting to "Functor" subconsciously. 
Thank you for accommodating by changing the name.

File: java/src/main/native/src/row_conversion.cu [hunk 10]
```diff
@@ -1199,28 +1134,27 @@ copy_validity_from_rows(const size_type num_rows, const size_type num_columns,
...
                                        size_type const num_string_columns) {
-  // Each warp takes a tile, which is a single column and up to ROWS_PER_BLOCK rows. A tile
-  // will not wrap around the bottom of the table. The warp will copy the strings for each row
-  // in the tile. Traversing in row-major order to coalesce the offsets and size reads.
+  // Each warp takes a tile, which is a single column and up to ROWS_PER_BLOCK rows. A tile will not
+  // wrap around the bottom of the table. The warp will copy the strings for each row in the tile.
+  // Traversing in row-major order to coalesce the offsets and size reads.
   auto my_block = cooperative_groups::this_thread_block();
-  auto my_partition = cooperative_groups::tiled_partition<32>(my_block);
+  auto warp = cooperative_groups::tiled_partition<cudf::detail::warp_size>(my_block);
```
Comment (mythrocks) [inline java/src/main/native/src/row_conversion.cu:1147]: :+1:

**Review by mythrocks — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]
LGTM :+1: 

My last recommendation is to move the `constexpr` values at the beginning of this file into an anonymous namespace if possible. Fine if not.

**Review by ttnghia — APPROVED** [review]

**Review by nvdbaranec — APPROVED** [review]

**Pull request merged**